collector/aws/resources/natgateway.go

```diff
@@ -0,0 +1,264 @@
+package resources
+
+import (
+	"errors"
+	"finala/collector"
+	"finala/collector/aws/common"
+	"finala/collector/aws/register"
+	"finala/collector/config"
+	"finala/expression"
+	"fmt"
+	"time"
+
+	awsClient "github.com/aws/aws-sdk-go/aws"
+	awsCloudwatch "github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/pricing"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// NatGatewayClientDescriptor is an interface defining the aws NAT gateway client
+type NatGatewayClientDescriptor interface {
+	DescribeNatGateways(*ec2.DescribeNatGatewaysInput) (*ec2.DescribeNatGatewaysOutput, error)
+}
+
+//NatGatewayManager describes NAT gateway struct
+type NatGatewayManager struct {
+	client             NatGatewayClientDescriptor
+	awsManager         common.AWSManager
+	namespace          string
+	servicePricingCode string
+	Name               collector.ResourceIdentifier
+}
+
+// DetectedNATGateway defines the detected AWS NAT gateways
+type DetectedNATGateway struct {
+	Region   string
+	Metric   string
+	SubnetID string
+	VPCID    string
+	collector.PriceDetectedFields
+}
+
+func init() {
+	register.Registry("natgateway", NewNATGatewayManager)
+}
+
+// NewNATGatewayManager implements AWS GO SDK for ec2 NAT gateway
+func NewNATGatewayManager(awsManager common.AWSManager, client interface{}) (common.ResourceDetection, error) {
+
+	if client == nil {
+		client = ec2.New(awsManager.GetSession())
+	}
+
+	natGatewayClient, ok := client.(NatGatewayClientDescriptor)
+	if !ok {
+		return nil, errors.New("invalid NAT gateway client")
+	}
+
+	return &NatGatewayManager{
+		client:             natGatewayClient,
+		awsManager:         awsManager,
+		namespace:          "AWS/NATGateway",
+		servicePricingCode: "AmazonEC2",
+		Name:               awsManager.GetResourceIdentifier("natgateway"),
+	}, nil
+}
+
+// Detect check with elasticache instance is under utilization
+func (ngw *NatGatewayManager) Detect(metrics []config.MetricConfig) (interface{}, error) {
+
+	log.WithFields(log.Fields{
+		"region":   ngw.awsManager.GetRegion(),
+		"resource": "natgateway",
+	}).Info("analyzing resource")
+
+	ngw.awsManager.GetCollector().CollectStart(ngw.Name)
+
+	DetectedNATGateways := []DetectedNATGateway{}
+
+	pricingRegionPrefix, err := ngw.awsManager.GetPricingClient().GetRegionPrefix(ngw.awsManager.GetRegion())
+	if err != nil {
+		log.WithError(err).WithFields(log.Fields{
+			"region": ngw.awsManager.GetRegion(),
+		}).Error("Could not get pricing region prefix")
+		ngw.awsManager.GetCollector().CollectError(ngw.Name, err)
+		return DetectedNATGateways, err
+	}
+
+	pricingUsageTypeFilter := []*pricing.Filter{
```

Comment:
Why are you not added this filter into `getPricingFilterInput`?

Reply:
good point , moved.